src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs

```diff
@@ -282,23 +218,121 @@ public override void WriteTo(TextWriter writer, HtmlEncoder encoder)
+            }
+        }
+
+        private void WriteToCore(object entry, TextWriter writer, HtmlEncoder encoder)
+        {
+            if (entry == null)
+            {
+                return;
+            }
+
+            var stringValue = entry as string;
+            if (stringValue != null)
+            {
+                encoder.Encode(writer, stringValue);
+            }
+            else
+            {
+                ((IHtmlContent)entry).WriteTo(writer, encoder);
+            }
+        }
+
+        private void CopyToCore(object entry, IHtmlContentBuilder destination)
+        {
+            if (entry == null)
+            {
+                return;
+            }
+
+            string entryAsString;
+            IHtmlContentContainer entryAsContainer;
+            if ((entryAsString = entry as string) != null)
+            {
+                destination.Append(entryAsString);
+            }
+            else if ((entryAsContainer = entry as IHtmlContentContainer) != null)
+            {
+                entryAsContainer.CopyTo(destination);
+            }
+            else
+            {
+                destination.AppendHtml((IHtmlContent)entry);
+            }
+        }
 
-                var stringValue = entry as string;
-                if (stringValue != null)
+        private void MoveToCore(object entry, IHtmlContentBuilder destination)
+        {
+            if (entry == null)
+            {
+                return;
+            }
+
+            string entryAsString;
+            IHtmlContentContainer entryAsContainer;
+            if ((entryAsString = entry as string) != null)
+            {
+                destination.Append(entryAsString);
+            }
+            else if ((entryAsContainer = entry as IHtmlContentContainer) != null)
+            {
+                entryAsContainer.MoveTo(destination);
+            }
+            else
+            {
+                destination.AppendHtml((IHtmlContent)entry);
+            }
+        }
+
+        private bool IsEmptyOrWhiteSpaceCore(object entry, EmptyOrWhiteSpaceWriter writer)
+        {
+            if (entry == null)
+            {
+                return false;
+            }
+
+            var stringValue = entry as string;
+            if (stringValue != null)
+            {
+                if (!string.IsNullOrWhiteSpace(stringValue))
                 {
-                    encoder.Encode(writer, stringValue);
+                    return false;
                 }
-                else
+            }
+            else
+            {
+                ((IHtmlContent)entry).WriteTo(writer, HtmlEncoder.Default);
+                if (!writer.IsEmptyOrWhiteSpace)
                 {
-                    ((IHtmlContent)entry).WriteTo(writer, encoder);
+                    return false;
                 }
             }
+
+            return true;
+        }
+
+        private TagHelperContent AppendCore(object entry)
+        {
+            if (!IsModified)
```

Comment:
This condition is incorrect and inefficient for `SetContent()` and similar cases -- will allocate buffers unecessarily. `IsModified` is `true` after `Clear()` is called but `Buffer` may remain `null`. Should instead have a clear `_isEmpty` field. The transitions should always be `_isEmpty` -> `_isSingleContentSet` -> `_buffer != null && _buffer.Count > 1` regardless of whether `IsModified` or `_buffer != null` initially i.e. after `Clear()` or `Reinitialize()` call.